// @cucumber/cucumber/lib/index.js
"use strict";
/**
 * User code functions and helpers
 *
 * @packageDocumentation
 * @module (root)
 * @remarks
 * These docs cover the functions and helpers for user code registration and test setup. The entry point is `@cucumber/cucumber`.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.wrapPromiseWithTimeout = exports.Status = exports.parallelCanAssignHelpers = exports.World = exports.HookTarget = exports.world = exports.context = exports.When = exports.Then = exports.setParallelCanAssign = exports.setWorldConstructor = exports.setDefinitionFunctionWrapper = exports.setDefaultTimeout = exports.Given = exports.defineParameterType = exports.defineStep = exports.BeforeStep = exports.BeforeAll = exports.Before = exports.AfterStep = exports.AfterAll = exports.After = exports.formatterHelpers = exports.version = exports.supportCodeLibraryBuilder = exports.TestCaseHookDefinition = exports.DataTable = exports.UsageJsonFormatter = exports.UsageFormatter = exports.SummaryFormatter = exports.SnippetsFormatter = exports.RerunFormatter = exports.ProgressFormatter = exports.JsonFormatter = exports.FormatterBuilder = exports.Formatter = void 0;
const messages_1 = require("@cucumber/messages");
const formatterHelpers = __importStar(require("./formatter/helpers"));
exports.formatterHelpers = formatterHelpers;
const support_code_library_builder_1 = __importDefault(require("./support_code_library_builder"));
const parallelCanAssignHelpers = __importStar(require("./support_code_library_builder/parallel_can_assign_helpers"));
exports.parallelCanAssignHelpers = parallelCanAssignHelpers;
// Formatters
var formatter_1 = require("./formatter");
Object.defineProperty(exports, "Formatter", { enumerable: true, get: function () { return __importDefault(formatter_1).default; } });
var builder_1 = require("./formatter/builder");
Object.defineProperty(exports, "FormatterBuilder", { enumerable: true, get: function () { return __importDefault(builder_1).default; } });
var json_formatter_1 = require("./formatter/json_formatter");
Object.defineProperty(exports, "JsonFormatter", { enumerable: true, get: function () { return __importDefault(json_formatter_1).default; } });
var progress_formatter_1 = require("./formatter/progress_formatter");
Object.defineProperty(exports, "ProgressFormatter", { enumerable: true, get: function () { return __importDefault(progress_formatter_1).default; } });
var rerun_formatter_1 = require("./formatter/rerun_formatter");
Object.defineProperty(exports, "RerunFormatter", { enumerable: true, get: function () { return __importDefault(rerun_formatter_1).default; } });
var snippets_formatter_1 = require("./formatter/snippets_formatter");
Object.defineProperty(exports, "SnippetsFormatter", { enumerable: true, get: function () { return __importDefault(snippets_formatter_1).default; } });
var summary_formatter_1 = require("./formatter/summary_formatter");
Object.defineProperty(exports, "SummaryFormatter", { enumerable: true, get: function () { return __importDefault(summary_formatter_1).default; } });
var usage_formatter_1 = require("./formatter/usage_formatter");
Object.defineProperty(exports, "UsageFormatter", { enumerable: true, get: function () { return __importDefault(usage_formatter_1).default; } });
var usage_json_formatter_1 = require("./formatter/usage_json_formatter");
Object.defineProperty(exports, "UsageJsonFormatter", { enumerable: true, get: function () { return __importDefault(usage_json_formatter_1).default; } });
var data_table_1 = require("./models/data_table");
Object.defineProperty(exports, "DataTable", { enumerable: true, get: function () { return __importDefault(data_table_1).default; } });
var test_case_hook_definition_1 = require("./models/test_case_hook_definition");
Object.defineProperty(exports, "TestCaseHookDefinition", { enumerable: true, get: function () { return __importDefault(test_case_hook_definition_1).default; } });
// Top level
var support_code_library_builder_2 = require("./support_code_library_builder");
Object.defineProperty(exports, "supportCodeLibraryBuilder", { enumerable: true, get: function () { return __importDefault(support_code_library_builder_2).default; } });
var version_1 = require("./version");
Object.defineProperty(exports, "version", { enumerable: true, get: function () { return version_1.version; } });
// Support Code Functions
const { methods } = support_code_library_builder_1.default;
exports.After = methods.After;
exports.AfterAll = methods.AfterAll;
exports.AfterStep = methods.AfterStep;
exports.Before = methods.Before;
exports.BeforeAll = methods.BeforeAll;
exports.BeforeStep = methods.BeforeStep;
exports.defineStep = methods.defineStep;
exports.defineParameterType = methods.defineParameterType;
exports.Given = methods.Given;
exports.setDefaultTimeout = methods.setDefaultTimeout;
exports.setDefinitionFunctionWrapper = methods.setDefinitionFunctionWrapper;
exports.setWorldConstructor = methods.setWorldConstructor;
exports.setParallelCanAssign = methods.setParallelCanAssign;
exports.Then = methods.Then;
exports.When = methods.When;
var scope_1 = require("./runtime/scope");
Object.defineProperty(exports, "context", { enumerable: true, get: function () { return scope_1.contextProxy; } });
Object.defineProperty(exports, "world", { enumerable: true, get: function () { return scope_1.worldProxy; } });
var types_1 = require("./support_code_library_builder/types");
Object.defineProperty(exports, "HookTarget", { enumerable: true, get: function () { return types_1.HookTarget; } });
var world_1 = require("./support_code_library_builder/world");
Object.defineProperty(exports, "World", { enumerable: true, get: function () { return __importDefault(world_1).default; } });
exports.Status = messages_1.TestStepResultStatus;
// Time helpers
var time_1 = require("./time");
Object.defineProperty(exports, "wrapPromiseWithTimeout", { enumerable: true, get: function () { return time_1.wrapPromiseWithTimeout; } });
