// @cucumber/cucumber/lib/api/emit_support_code_messages.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.emitMetaMessage = emitMetaMessage;
exports.emitSupportCodeMessages = emitSupportCodeMessages;
const node_os_1 = __importDefault(require("node:os"));
const ci_environment_1 = __importDefault(require("@cucumber/ci-environment"));
const messages_1 = require("@cucumber/messages");
const version_1 = require("../version");
async function emitMetaMessage(eventBroadcaster, env) {
    const meta = {
        protocolVersion: messages_1.version,
        implementation: {
            version: version_1.version,
            name: 'cucumber-js',
        },
        cpu: {
            name: node_os_1.default.arch(),
        },
        os: {
            name: node_os_1.default.platform(),
            version: node_os_1.default.release(),
        },
        runtime: {
            name: 'node.js',
            version: process.versions.node,
        },
        ci: (0, ci_environment_1.default)(env),
    };
    eventBroadcaster.emit('envelope', {
        meta,
    });
}
function makeSourceReference(source) {
    return {
        uri: source.uri,
        location: {
            line: source.line,
        },
    };
}
function extractPatternSource(pattern) {
    if (pattern instanceof RegExp) {
        return pattern.flags ? pattern.toString() : pattern.source;
    }
    return pattern;
}
function collectParameterTypeEnvelopes(supportCodeLibrary, newId) {
    const ordered = [];
    for (const parameterType of supportCodeLibrary.parameterTypeRegistry.parameterTypes) {
        if (parameterType.builtin) {
            continue;
        }
        const source = supportCodeLibrary.parameterTypeRegistry.lookupSource(parameterType);
        ordered.push({
            order: source.order,
            envelope: {
                parameterType: {
                    id: newId(),
                    name: parameterType.name,
                    preferForRegularExpressionMatch: parameterType.preferForRegexpMatch,
                    regularExpressions: parameterType.regexpStrings,
                    useForSnippets: parameterType.useForSnippets,
                    sourceReference: makeSourceReference(source),
                },
            },
        });
    }
    return ordered;
}
function collectStepDefinitionEnvelopes(supportCodeLibrary) {
    return supportCodeLibrary.stepDefinitions.map((stepDefinition) => ({
        order: stepDefinition.order,
        envelope: {
            stepDefinition: {
                id: stepDefinition.id,
                pattern: {
                    source: extractPatternSource(stepDefinition.pattern),
                    type: typeof stepDefinition.pattern === 'string'
                        ? messages_1.StepDefinitionPatternType.CUCUMBER_EXPRESSION
                        : messages_1.StepDefinitionPatternType.REGULAR_EXPRESSION,
                },
                sourceReference: makeSourceReference(stepDefinition),
            },
        },
    }));
}
function collectHookEnvelopes(supportCodeLibrary) {
    const allHooks = [
        [supportCodeLibrary.beforeTestCaseHookDefinitions, messages_1.HookType.BEFORE_TEST_CASE],
        [supportCodeLibrary.afterTestCaseHookDefinitions, messages_1.HookType.AFTER_TEST_CASE],
        [supportCodeLibrary.beforeTestRunHookDefinitions, messages_1.HookType.BEFORE_TEST_RUN],
        [supportCodeLibrary.afterTestRunHookDefinitions, messages_1.HookType.AFTER_TEST_RUN],
    ];
    const ordered = [];
    allHooks.forEach(([hooks, type]) => {
        hooks.forEach((hook) => {
            ordered.push({
                order: hook.order,
                envelope: {
                    hook: {
                        id: hook.id,
                        type,
                        name: hook.name,
                        ...('tagExpression' in hook && {
                            tagExpression: hook.tagExpression,
                        }),
                        sourceReference: makeSourceReference(hook),
                    },
                },
            });
        });
    });
    return ordered;
}
function emitSupportCodeMessages({ eventBroadcaster, supportCodeLibrary, newId, }) {
    const orderedEnvelopes = [
        ...collectParameterTypeEnvelopes(supportCodeLibrary, newId),
        ...collectStepDefinitionEnvelopes(supportCodeLibrary),
        ...collectHookEnvelopes(supportCodeLibrary),
    ];
    for (const { envelope } of orderedEnvelopes.sort((a, b) => a.order - b.order)) {
        eventBroadcaster.emit('envelope', envelope);
    }
    for (const undefinedParameterType of supportCodeLibrary.undefinedParameterTypes) {
        eventBroadcaster.emit('envelope', { undefinedParameterType });
    }
}
