// @cucumber/cucumber/lib/formatter/rerun_formatter.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const messages_1 = require("@cucumber/messages");
const value_checker_1 = require("../value_checker");
const _1 = __importDefault(require("./"));
const gherkin_document_parser_1 = require("./helpers/gherkin_document_parser");
const DEFAULT_SEPARATOR = '\n';
function isFailedAttempt(worstTestStepResult) {
    return worstTestStepResult.status !== messages_1.TestStepResultStatus.PASSED;
}
class RerunFormatter extends _1.default {
    separator;
    static documentation = 'Prints failing files with line numbers.';
    constructor(options) {
        super(options);
        options.eventBroadcaster.on('envelope', (envelope) => {
            if ((0, value_checker_1.doesHaveValue)(envelope.testRunFinished)) {
                this.logFailedTestCases();
            }
        });
        const rerunOptions = (0, value_checker_1.valueOrDefault)(options.parsedArgvOptions.rerun, {});
        this.separator = (0, value_checker_1.valueOrDefault)(rerunOptions.separator, DEFAULT_SEPARATOR);
    }
    getFailureMap() {
        const mapping = {};
        this.eventDataCollector
            .getTestCaseAttempts()
            .forEach(({ gherkinDocument, pickle, worstTestStepResult, willBeRetried }) => {
            if (isFailedAttempt(worstTestStepResult) && !willBeRetried) {
                const relativeUri = pickle.uri;
                const line = (0, gherkin_document_parser_1.getGherkinScenarioLocationMap)(gherkinDocument)[pickle.astNodeIds[pickle.astNodeIds.length - 1]].line;
                if ((0, value_checker_1.doesNotHaveValue)(mapping[relativeUri])) {
                    mapping[relativeUri] = [];
                }
                mapping[relativeUri].push(line);
            }
        });
        return mapping;
    }
    formatFailedTestCases() {
        const mapping = this.getFailureMap();
        return Object.keys(mapping)
            .map((uri) => {
            const lines = mapping[uri];
            return `${uri}:${lines.join(':')}`;
        })
            .join(this.separator);
    }
    logFailedTestCases() {
        const failedTestCases = this.formatFailedTestCases();
        this.log(failedTestCases);
    }
}
exports.default = RerunFormatter;
