// @cucumber/cucumber/lib/runtime/executor.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Executor = void 0;
const messages_1 = require("@cucumber/messages");
const user_code_runner_1 = __importDefault(require("../user_code_runner"));
const value_checker_1 = require("../value_checker");
const format_error_1 = require("./format_error");
const helpers_1 = require("./helpers");
const scope_1 = require("./scope");
const stopwatch_1 = require("./stopwatch");
const test_case_runner_1 = __importDefault(require("./test_case_runner"));
/**
 * Runs individual units of work asynchronously
 * @remarks
 * Unaware of the overall test run scope, or whether it's on the main thread,
 * or whether any other executors are in play.
 */
class Executor {
    testRunStartedId;
    workerId;
    eventBroadcaster;
    newId;
    options;
    supportCodeLibrary;
    snippetBuilder;
    constructor(testRunStartedId, workerId, eventBroadcaster, newId, options, supportCodeLibrary, snippetBuilder) {
        this.testRunStartedId = testRunStartedId;
        this.workerId = workerId;
        this.eventBroadcaster = eventBroadcaster;
        this.newId = newId;
        this.options = options;
        this.supportCodeLibrary = supportCodeLibrary;
        this.snippetBuilder = snippetBuilder;
    }
    async runTestRunHook(hookDefinition) {
        const testRunHookStartedId = this.newId();
        this.eventBroadcaster.emit('envelope', {
            testRunHookStarted: {
                testRunStartedId: this.testRunStartedId,
                workerId: this.workerId,
                id: testRunHookStartedId,
                hookId: hookDefinition.id,
                timestamp: (0, stopwatch_1.timestamp)(),
            },
        });
        let result;
        if (this.options.dryRun) {
            result = {
                duration: {
                    seconds: 0,
                    nanos: 0,
                },
                status: messages_1.TestStepResultStatus.SKIPPED,
            };
        }
        else {
            const stopwatch = (0, stopwatch_1.create)().start();
            const context = { parameters: this.options.worldParameters };
            const { error } = await (0, scope_1.runInTestRunScope)({ context }, () => user_code_runner_1.default.run({
                argsArray: [],
                fn: hookDefinition.code,
                thisArg: context,
                timeoutInMilliseconds: (0, value_checker_1.valueOrDefault)(hookDefinition.options.timeout, this.supportCodeLibrary.defaultTimeout),
            }));
            const duration = stopwatch.stop().duration();
            if ((0, value_checker_1.doesHaveValue)(error)) {
                result = {
                    duration,
                    status: messages_1.TestStepResultStatus.FAILED,
                    ...(0, format_error_1.formatError)(error, this.options.filterStacktraces),
                };
            }
            else {
                result = {
                    duration,
                    status: messages_1.TestStepResultStatus.PASSED,
                };
            }
        }
        this.eventBroadcaster.emit('envelope', {
            testRunHookFinished: {
                testRunHookStartedId,
                result,
                timestamp: (0, stopwatch_1.timestamp)(),
            },
        });
        return result.status !== messages_1.TestStepResultStatus.FAILED;
    }
    async runBeforeAllHooks(predicate = () => true) {
        let success = true;
        for (const hookDefinition of this.supportCodeLibrary.beforeTestRunHookDefinitions) {
            if (!predicate(hookDefinition)) {
                continue;
            }
            if (!(await this.runTestRunHook(hookDefinition))) {
                success = false;
            }
        }
        return success;
    }
    async runTestCase({ gherkinDocument, pickle, testCase }, failing) {
        const testCaseRunner = new test_case_runner_1.default({
            workerId: this.workerId,
            eventBroadcaster: this.eventBroadcaster,
            newId: this.newId,
            gherkinDocument,
            pickle,
            testCase,
            retries: (0, helpers_1.retriesForPickle)(pickle, this.options),
            skip: this.options.dryRun || (this.options.failFast && failing),
            filterStackTraces: this.options.filterStacktraces,
            supportCodeLibrary: this.supportCodeLibrary,
            worldParameters: this.options.worldParameters,
            snippetBuilder: this.snippetBuilder,
        });
        const status = await testCaseRunner.run();
        return !(0, helpers_1.shouldCauseFailure)(status, this.options);
    }
    async runAfterAllHooks(predicate = () => true) {
        let success = true;
        const reversed = [...this.supportCodeLibrary.afterTestRunHookDefinitions].reverse();
        for (const hookDefinition of reversed) {
            if (!predicate(hookDefinition)) {
                continue;
            }
            if (!(await this.runTestRunHook(hookDefinition))) {
                success = false;
            }
        }
        return success;
    }
}
exports.Executor = Executor;
