// @cucumber/cucumber/lib/runtime/helpers.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAmbiguousStepException = getAmbiguousStepException;
exports.retriesForPickle = retriesForPickle;
exports.shouldCauseFailure = shouldCauseFailure;
const messages_1 = require("@cucumber/messages");
const cli_table3_1 = __importDefault(require("cli-table3"));
const indent_string_1 = __importDefault(require("indent-string"));
const location_helpers_1 = require("../formatter/helpers/location_helpers");
const pickle_filter_1 = require("../pickle_filter");
function getAmbiguousStepException(stepDefinitions) {
    const table = new cli_table3_1.default({
        chars: {
            bottom: '',
            'bottom-left': '',
            'bottom-mid': '',
            'bottom-right': '',
            left: '',
            'left-mid': '',
            mid: '',
            'mid-mid': '',
            middle: ' - ',
            right: '',
            'right-mid': '',
            top: '',
            'top-left': '',
            'top-mid': '',
            'top-right': '',
        },
        style: {
            border: [],
            'padding-left': 0,
            'padding-right': 0,
        },
    });
    table.push(...stepDefinitions.map((stepDefinition) => {
        const pattern = stepDefinition.pattern.toString();
        return [pattern, (0, location_helpers_1.formatLocation)(stepDefinition)];
    }));
    return `${'Multiple step definitions match:' + '\n'}${(0, indent_string_1.default)(table.toString(), 2)}`;
}
function retriesForPickle(pickle, options) {
    if (!options.retry) {
        return 0;
    }
    const retries = options.retry;
    if (retries === 0) {
        return 0;
    }
    const retryTagFilter = options.retryTagFilter;
    if (!retryTagFilter) {
        return retries;
    }
    const pickleTagFilter = new pickle_filter_1.PickleTagFilter(retryTagFilter);
    if (pickleTagFilter.matchesAllTagExpressions(pickle)) {
        return retries;
    }
    return 0;
}
function shouldCauseFailure(status, options) {
    if (options.dryRun) {
        return false;
    }
    const failureStatuses = [
        messages_1.TestStepResultStatus.AMBIGUOUS,
        messages_1.TestStepResultStatus.FAILED,
        messages_1.TestStepResultStatus.UNDEFINED,
    ];
    if (options.strict) {
        failureStatuses.push(messages_1.TestStepResultStatus.PENDING);
    }
    return failureStatuses.includes(status);
}
