// @cucumber/cucumber/lib/runtime/attachment_manager/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const node_stream_1 = require("node:stream");
const messages_1 = require("@cucumber/messages");
const value_checker_1 = require("../../value_checker");
class AttachmentManager {
    onAttachment;
    constructor(onAttachment) {
        this.onAttachment = onAttachment;
    }
    log(text) {
        return this.create(text, 'text/x.cucumber.log+plain');
    }
    link(...url) {
        return this.create(url.join('\n'), 'text/uri-list');
    }
    create(data, mediaTypeOrOptions, callback) {
        const options = normaliseOptions(mediaTypeOrOptions);
        if (Buffer.isBuffer(data)) {
            if ((0, value_checker_1.doesNotHaveValue)(options.mediaType)) {
                throw Error('Buffer attachments must specify a media type');
            }
            this.createBufferAttachment(data, options.mediaType, options.fileName);
        }
        else if (typeof data === 'string') {
            if ((0, value_checker_1.doesNotHaveValue)(options.mediaType)) {
                options.mediaType = 'text/plain';
            }
            if (options.mediaType.startsWith('base64:')) {
                this.createStringAttachment(data, {
                    encoding: messages_1.AttachmentContentEncoding.BASE64,
                    contentType: options.mediaType.replace('base64:', ''),
                }, options.fileName);
            }
            else {
                this.createStringAttachment(data, {
                    encoding: messages_1.AttachmentContentEncoding.IDENTITY,
                    contentType: options.mediaType,
                }, options.fileName);
            }
        }
        else if ((0, node_stream_1.isReadable)(data)) {
            if ((0, value_checker_1.doesNotHaveValue)(options.mediaType)) {
                throw Error('Stream attachments must specify a media type');
            }
            return this.createStreamAttachment(data, options.mediaType, options.fileName, callback);
        }
        else {
            throw Error('Invalid attachment data: must be a buffer, readable stream, or string');
        }
    }
    createBufferAttachment(data, mediaType, fileName) {
        this.createStringAttachment(data.toString('base64'), {
            encoding: messages_1.AttachmentContentEncoding.BASE64,
            contentType: mediaType,
        }, fileName);
    }
    createStreamAttachment(data, mediaType, fileName, callback) {
        const promise = new Promise((resolve, reject) => {
            const buffers = [];
            data.on('data', (chunk) => {
                buffers.push(chunk);
            });
            data.on('end', () => {
                this.createBufferAttachment(Buffer.concat(buffers), mediaType, fileName);
                resolve();
            });
            data.on('error', reject);
        });
        if ((0, value_checker_1.doesHaveValue)(callback)) {
            promise.then(callback, callback);
        }
        else {
            return promise;
        }
    }
    createStringAttachment(data, media, fileName) {
        this.onAttachment({
            data,
            media,
            ...(fileName ? { fileName } : {}),
        });
    }
}
exports.default = AttachmentManager;
function normaliseOptions(mediaTypeOrOptions) {
    if (!mediaTypeOrOptions) {
        return {};
    }
    if (typeof mediaTypeOrOptions === 'string') {
        return {
            mediaType: mediaTypeOrOptions,
        };
    }
    return mediaTypeOrOptions;
}
