// @cucumber/cucumber/lib/formatter/helpers/event_data_collector.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const messages_1 = require("@cucumber/messages");
const value_checker_1 = require("../../value_checker");
class EventDataCollector {
    gherkinDocumentMap = {};
    pickleMap = {};
    testCaseMap = {};
    testCaseAttemptDataMap = {};
    undefinedParameterTypes = [];
    constructor(eventBroadcaster) {
        eventBroadcaster.on('envelope', this.parseEnvelope.bind(this));
    }
    getGherkinDocument(uri) {
        return this.gherkinDocumentMap[uri];
    }
    getPickle(pickleId) {
        return this.pickleMap[pickleId];
    }
    getTestCaseAttempts() {
        return Object.keys(this.testCaseAttemptDataMap).map((testCaseStartedId) => {
            return this.getTestCaseAttempt(testCaseStartedId);
        });
    }
    getTestCaseAttempt(testCaseStartedId) {
        const testCaseAttemptData = this.testCaseAttemptDataMap[testCaseStartedId];
        const testCase = this.testCaseMap[testCaseAttemptData.testCaseId];
        const pickle = this.pickleMap[testCase.pickleId];
        return {
            gherkinDocument: this.gherkinDocumentMap[pickle.uri],
            pickle,
            testCase,
            attempt: testCaseAttemptData.attempt,
            willBeRetried: testCaseAttemptData.willBeRetried,
            stepAttachments: testCaseAttemptData.stepAttachments,
            stepResults: testCaseAttemptData.stepResults,
            worstTestStepResult: testCaseAttemptData.worstTestStepResult,
        };
    }
    parseEnvelope(envelope) {
        if ((0, value_checker_1.doesHaveValue)(envelope.gherkinDocument)) {
            this.gherkinDocumentMap[envelope.gherkinDocument.uri] = envelope.gherkinDocument;
        }
        else if ((0, value_checker_1.doesHaveValue)(envelope.pickle)) {
            this.pickleMap[envelope.pickle.id] = envelope.pickle;
        }
        else if ((0, value_checker_1.doesHaveValue)(envelope.undefinedParameterType)) {
            this.undefinedParameterTypes.push(envelope.undefinedParameterType);
        }
        else if ((0, value_checker_1.doesHaveValue)(envelope.testCase)) {
            this.testCaseMap[envelope.testCase.id] = envelope.testCase;
        }
        else if ((0, value_checker_1.doesHaveValue)(envelope.testCaseStarted)) {
            this.initTestCaseAttempt(envelope.testCaseStarted);
        }
        else if ((0, value_checker_1.doesHaveValue)(envelope.attachment)) {
            this.storeAttachment(envelope.attachment);
        }
        else if ((0, value_checker_1.doesHaveValue)(envelope.testStepFinished)) {
            this.storeTestStepResult(envelope.testStepFinished);
        }
        else if ((0, value_checker_1.doesHaveValue)(envelope.testCaseFinished)) {
            this.storeTestCaseResult(envelope.testCaseFinished);
        }
    }
    initTestCaseAttempt(testCaseStarted) {
        // pre-seed a fallback UNKNOWN result for every step
        const stepResults = {};
        for (const testStep of this.testCaseMap[testCaseStarted.testCaseId].testSteps) {
            stepResults[testStep.id] = {
                duration: { seconds: 0, nanos: 0 },
                status: messages_1.TestStepResultStatus.UNKNOWN,
            };
        }
        this.testCaseAttemptDataMap[testCaseStarted.id] = {
            attempt: testCaseStarted.attempt,
            willBeRetried: false,
            testCaseId: testCaseStarted.testCaseId,
            stepAttachments: {},
            stepResults,
            worstTestStepResult: {
                duration: { seconds: 0, nanos: 0 },
                status: messages_1.TestStepResultStatus.UNKNOWN,
            },
        };
    }
    storeAttachment(attachment) {
        const { testCaseStartedId, testStepId } = attachment;
        if ((0, value_checker_1.doesHaveValue)(testCaseStartedId) && (0, value_checker_1.doesHaveValue)(testStepId)) {
            const { stepAttachments } = this.testCaseAttemptDataMap[testCaseStartedId];
            if ((0, value_checker_1.doesNotHaveValue)(stepAttachments[testStepId])) {
                stepAttachments[testStepId] = [];
            }
            stepAttachments[testStepId].push(attachment);
        }
    }
    storeTestStepResult({ testCaseStartedId, testStepId, testStepResult }) {
        this.testCaseAttemptDataMap[testCaseStartedId].stepResults[testStepId] = testStepResult;
    }
    storeTestCaseResult({ testCaseStartedId, willBeRetried }) {
        const stepResults = Object.values(this.testCaseAttemptDataMap[testCaseStartedId].stepResults);
        this.testCaseAttemptDataMap[testCaseStartedId].worstTestStepResult =
            (0, messages_1.getWorstTestStepResult)(stepResults);
        this.testCaseAttemptDataMap[testCaseStartedId].willBeRetried = willBeRetried;
    }
}
exports.default = EventDataCollector;
