// @cucumber/cucumber/lib/runtime/serial/adapter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InProcessAdapter = void 0;
const executor_1 = require("../executor");
/**
 * A simple adapter that executes all work in serial on the main thread
 */
class InProcessAdapter {
    executor;
    constructor(testRunStartedId, eventBroadcaster, newId, options, supportCodeLibrary, snippetBuilder) {
        this.executor = new executor_1.Executor(testRunStartedId, undefined, eventBroadcaster, newId, options, supportCodeLibrary, snippetBuilder);
    }
    async setup() {
        // no-op for serial runtime
    }
    async teardown() {
        // no-op for serial runtime
    }
    async runBeforeAllHooks() {
        return await this.executor.runBeforeAllHooks();
    }
    async runTestCases(assembledTestCases) {
        let failing = false;
        for (const item of assembledTestCases) {
            if (!(await this.executor.runTestCase(item, failing))) {
                failing = true;
            }
        }
        return !failing;
    }
    async runAfterAllHooks() {
        return await this.executor.runAfterAllHooks();
    }
}
exports.InProcessAdapter = InProcessAdapter;
