// @cucumber/cucumber/lib/runtime/parallel/adapter.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorkerThreadsAdapter = void 0;
const node_path_1 = __importDefault(require("node:path"));
const node_worker_threads_1 = require("node:worker_threads");
const types_1 = require("../../support_code_library_builder/types");
const executor_1 = require("../executor");
const test_cases_phase_1 = require("./test_cases_phase");
const test_run_hooks_phase_1 = require("./test_run_hooks_phase");
/**
 * An adapter that distributes work across multiple worker threads
 * @remarks
 * Each phase of the test run is self-contained and self-orchestrating - every
 * FINISHED message from a worker may cause the next piece of work to be
 * triggered or the phase to be settled.
 */
class WorkerThreadsAdapter {
    testRunStartedId;
    environment;
    logger;
    eventBroadcaster;
    options;
    snippetOptions;
    supportCodeLibrary;
    readiness;
    phase;
    tearingDown = false;
    workers = new Set();
    running = new Map();
    executor;
    constructor(testRunStartedId, environment, logger, eventBroadcaster, newId, options, snippetOptions, supportCodeLibrary, snippetBuilder) {
        this.testRunStartedId = testRunStartedId;
        this.environment = environment;
        this.logger = logger;
        this.eventBroadcaster = eventBroadcaster;
        this.options = options;
        this.snippetOptions = snippetOptions;
        this.supportCodeLibrary = supportCodeLibrary;
        this.executor = new executor_1.Executor(testRunStartedId, undefined, eventBroadcaster, newId, options, supportCodeLibrary, snippetBuilder);
    }
    async setup() {
        await new Promise((resolve, reject) => {
            this.readiness = { resolve, reject };
            const total = this.options.parallel;
            for (let i = 0; i < total; i++) {
                const id = i.toString();
                // spin up a dedicated message channel for coordinator-worker comms
                const { port1, port2 } = new node_worker_threads_1.MessageChannel();
                const workerThread = new node_worker_threads_1.Worker(node_path_1.default.resolve(__dirname, 'worker.mjs'), {
                    env: {
                        ...this.environment.env,
                        CUCUMBER_PARALLEL: 'true',
                        CUCUMBER_TOTAL_WORKERS: total.toString(),
                        CUCUMBER_WORKER_ID: id,
                    },
                    resourceLimits: this.options.workerOptions?.resourceLimits,
                    workerData: {
                        cwd: this.environment.cwd,
                        testRunStartedId: this.testRunStartedId,
                        supportCodeCoordinates: this.supportCodeLibrary.originalCoordinates,
                        supportCodeIds: {
                            stepDefinitionIds: this.supportCodeLibrary.stepDefinitions.map((s) => s.id),
                            beforeTestCaseHookDefinitionIds: this.supportCodeLibrary.beforeTestCaseHookDefinitions.map((h) => h.id),
                            afterTestCaseHookDefinitionIds: this.supportCodeLibrary.afterTestCaseHookDefinitions.map((h) => h.id),
                            beforeTestRunHookDefinitionIds: this.supportCodeLibrary.beforeTestRunHookDefinitions.map((h) => h.id),
                            afterTestRunHookDefinitionIds: this.supportCodeLibrary.afterTestRunHookDefinitions.map((h) => h.id),
                        },
                        options: this.options,
                        snippetOptions: this.snippetOptions,
                        port: port2,
                    },
                    transferList: [port2],
                });
                const worker = {
                    id,
                    workerThread,
                    port: port1,
                    ready: false,
                };
                this.workers.add(worker);
                port1.on('message', (event) => {
                    this.handleEventFromWorker(worker, event);
                });
                workerThread.on('error', (error) => {
                    this.handleErrorFromWorker(error, worker);
                });
                workerThread.on('exit', (exitCode) => {
                    this.handleExitFromWorker(exitCode, worker);
                });
            }
        });
        delete this.readiness;
    }
    async runBeforeAllHooks() {
        const coordinatorSuccess = await this.executor.runBeforeAllHooks((hook) => hook.on === types_1.HookTarget.COORDINATOR);
        const workersSuccess = await new Promise((resolve, reject) => {
            this.phase = new test_run_hooks_phase_1.TestRunHooksPhase(resolve, reject, 'BEFOREALL_HOOKS');
            this.startPhase();
        });
        delete this.phase;
        return coordinatorSuccess && workersSuccess;
    }
    async runTestCases(assembledTestCases) {
        const success = await new Promise((resolve, reject) => {
            this.phase = new test_cases_phase_1.TestCasesPhase(resolve, reject, this.logger, this.supportCodeLibrary.parallelCanAssign, assembledTestCases);
            this.startPhase();
        });
        delete this.phase;
        return success;
    }
    async runAfterAllHooks() {
        const workersSuccess = await new Promise((resolve, reject) => {
            this.phase = new test_run_hooks_phase_1.TestRunHooksPhase(resolve, reject, 'AFTERALL_HOOKS');
            this.startPhase();
        });
        delete this.phase;
        const coordinatorSuccess = await this.executor.runAfterAllHooks((hook) => hook.on === types_1.HookTarget.COORDINATOR);
        return coordinatorSuccess && workersSuccess;
    }
    async teardown() {
        this.tearingDown = true;
        for (const worker of this.workers.values()) {
            await worker.workerThread.terminate();
            // close our end of the channel so it stops keeping the loop alive
            worker.port.close();
        }
    }
    startPhase() {
        for (const worker of this.workers) {
            const command = this.phase.fill();
            if (command) {
                this.issueCommandToWorker(worker, command);
            }
        }
    }
    issueCommandToWorker(worker, command) {
        this.running.set(worker, command);
        worker.port.postMessage(command);
    }
    handleEventFromWorker(worker, event) {
        switch (event.type) {
            case 'READY':
                worker.ready = true;
                if ([...this.workers].every((mw) => mw.ready)) {
                    this.readiness?.resolve();
                }
                break;
            case 'ENVELOPE':
                this.eventBroadcaster.emit('envelope', event.envelope);
                break;
            case 'FINISHED': {
                const previousCommand = this.running.get(worker);
                this.running.delete(worker);
                const nextCommand = this.phase?.next(previousCommand, event);
                if (nextCommand) {
                    this.issueCommandToWorker(worker, nextCommand);
                }
                break;
            }
        }
    }
    handleErrorFromWorker(error, worker) {
        this.fail(new Error(`Error on worker ${worker.id}`, { cause: error }));
    }
    handleExitFromWorker(exitCode, worker) {
        if (!this.tearingDown) {
            this.fail(new Error(`Worker ${worker.id} exited unexpectedly with code ${exitCode}`));
        }
    }
    fail(reason) {
        this.readiness?.reject(reason);
        this.phase?.reject(reason);
        void this.teardown();
    }
}
exports.WorkerThreadsAdapter = WorkerThreadsAdapter;
