// @cucumber/gherkin-utils/dist/walkGherkinDocument.js
/**
 * Walks a Gherkin Document, visiting each node depth first (in the order they appear in the source)
 *
 * @param gherkinDocument
 * @param initialValue the initial value of the traversal
 * @param handlers handlers for each node type, which may return a new value
 * @return result the final value
 */
export function walkGherkinDocument(gherkinDocument, initialValue, handlers) {
    const commentsStack = gherkinDocument.comments.slice();
    let acc = initialValue;
    const h = { ...makeDefaultHandlers(), ...handlers };
    const feature = gherkinDocument.feature;
    acc = walkComments(popCommentsUntil(feature?.location), acc);
    if (!feature) {
        return acc;
    }
    acc = walkTags(feature.tags || [], acc);
    acc = h.feature(feature, acc);
    for (const child of feature.children) {
        if (child.background) {
            acc = walkStepContainer(child.background, acc);
        }
        else if (child.scenario) {
            acc = walkStepContainer(child.scenario, acc);
        }
        else if (child.rule) {
            acc = walkTags(child.rule.tags || [], acc);
            acc = walkComments(popCommentsUntil(child.rule.location), acc);
            acc = h.rule(child.rule, acc);
            for (const ruleChild of child.rule.children) {
                if (ruleChild.background) {
                    acc = walkStepContainer(ruleChild.background, acc);
                }
                else if (ruleChild.scenario) {
                    acc = walkStepContainer(ruleChild.scenario, acc);
                }
            }
        }
    }
    acc = walkComments(popRemainingContents(), acc);
    return acc;
    function walkComments(comments, acc) {
        return comments.reduce((acc, comment) => h.comment(comment, acc), acc);
    }
    function walkTags(tags, acc) {
        return tags.reduce((acc, tag) => h.tag(tag, acc), acc);
    }
    function walkSteps(steps, acc) {
        return steps.reduce((acc, step) => walkStep(step, acc), acc);
    }
    function walkStep(step, acc) {
        acc = walkComments(popCommentsUntil(step.location), acc);
        acc = h.step(step, acc);
        if (step.docString) {
            acc = h.docString(step.docString, acc);
        }
        if (step.dataTable) {
            acc = h.dataTable(step.dataTable, acc);
            acc = walkTableRows(step.dataTable.rows, acc);
        }
        return acc;
    }
    function walkTableRows(tableRows, acc) {
        return tableRows.reduce((acc, tableRow) => walkTableRow(tableRow, acc), acc);
    }
    function walkTableRow(tableRow, acc) {
        acc = h.tableRow(tableRow, acc);
        return tableRow.cells.reduce((acc, tableCell) => h.tableCell(tableCell, acc), acc);
    }
    function walkStepContainer(stepContainer, acc) {
        acc = walkComments(popCommentsUntil(stepContainer.location), acc);
        const scenario = 'tags' in stepContainer ? stepContainer : null;
        acc = walkTags(scenario?.tags || [], acc);
        acc = scenario ? h.scenario(scenario, acc) : h.background(stepContainer, acc);
        acc = walkSteps(stepContainer.steps, acc);
        if (scenario) {
            for (const examples of scenario.examples || []) {
                acc = walkComments(popCommentsUntil(examples.location), acc);
                acc = walkTags(examples.tags || [], acc);
                acc = h.examples(examples, acc);
                if (examples.tableHeader) {
                    acc = walkTableRow(examples.tableHeader, acc);
                    acc = walkTableRows(examples.tableBody || [], acc);
                }
            }
        }
        return acc;
    }
    function popCommentsUntil(location) {
        let count = 0;
        for (const comment of commentsStack) {
            if (location === undefined || comment.location.line < location.line) {
                count++;
            }
            else {
                break;
            }
        }
        return commentsStack.splice(0, count);
    }
    function popRemainingContents() {
        return commentsStack.splice(0, commentsStack.length);
    }
}
function makeDefaultHandlers() {
    const defaultHandlers = {
        feature(_feature, acc) {
            return acc;
        },
        background(_background, acc) {
            return acc;
        },
        rule(_rule, acc) {
            return acc;
        },
        scenario(_scenario, acc) {
            return acc;
        },
        step(_step, acc) {
            return acc;
        },
        examples(_examples, acc) {
            return acc;
        },
        tag(_tag, acc) {
            return acc;
        },
        comment(_comment, acc) {
            return acc;
        },
        dataTable(_dataTable, acc) {
            return acc;
        },
        tableRow(_tableRow, acc) {
            return acc;
        },
        tableCell(_tableCell, acc) {
            return acc;
        },
        docString(_docString, acc) {
            return acc;
        },
    };
    return defaultHandlers;
}
